feat: Add Arc's boosts implementation, b=no-bug, c=boosts

FlorianButz opened the pull request:
This PR aims to implement a basic version of the Boosts feature from Arc. This currently features font customization, theme customization and smart inverting.

https://github.com/user-attachments/assets/e092bacd-f3d2-41b9-ae84-a7b9a19d3e8f

FlorianButz committed:
Add basic boosts implementation

12th-devs commented:
I love this so much!

mr-cheffy committed:
feat: Add boosts backend boilerplate, b=no-bug, c=no-component

mr-cheffy commented:
I've added the boilerplate code we should use in order to change how firefox renders colors for certain shells. If you could make it in a window instead of a popup, it would be great via using `Services.obs` to communicate. Would also remove the need for ZenBoosts.mjs.

mr-cheffy commented:
Very excited for the implementation I came up with!

We edit firefox's rendering engine to "filter" the colors that where going to be displayed on the screen, unlike arc which I believe they use filters on the document. This means our solution is more efficient + more compatible with websites

Example:

<img width="3288" height="1944" alt="image" src="https://github.com/user-attachments/assets/fc86b090-8fb4-4f0f-9d13-a7d4fff59c7a" />


And here's where zen's (1) vs arc's (2) implementation really differ:

zen:
<img width="3368" height="2018" alt="image" src="https://github.com/user-attachments/assets/f1e72e38-6527-4dc0-b35a-63903b6e64b2" />

arc:
<img width="2652" height="1756" alt="image" src="https://github.com/user-attachments/assets/8b67e66c-c45c-4edb-8c1b-0015a2e83f59" />

mr-cheffy committed:
Merge branch 'dev' into dev

mr-cheffy committed:
feat: Add changes to firefox engine, b=no-bug, c=no-component

mr-cheffy committed:
Merge branch 'dev' of https://github.com/FlorianButz/desktop into FlorianButz/dev

mr-cheffy committed:
chore: Fix builds and lint, b=no-bug, c=no-component

FlorianButz committed:
Change boost editor to be a separate window

Remove unecessary ZenBoosts.mjs
Update ZenBoostsEditor to be a module

FlorianButz committed:
Add check if current tab can be boosted

FlorianButz committed:
Run lint:fix

mr-cheffy committed:
feat: Expose the c++ API to JS, b=no-bug, c=no-component

mr-cheffy committed:
Merge branch 'dev' of https://github.com/FlorianButz/desktop into FlorianButz/dev

FlorianButz committed:
Add boost panel in site specific settings

Add animated sparkle icon to control center if site is boosted
Fix bad management with the boost/registered css sheet maps
Remove boost-editor.inc and references

FlorianButz committed:
Merge branch 'dev' of https://github.com/FlorianButz/desktop into dev

mr-cheffy committed:
Merge branch 'dev' into dev

mr-cheffy committed:
feat: Make boost data be stored on the browsing context, b=no-bug, c=glance

FlorianButz committed:
chore: use better icons

mr-cheffy committed:
feat: Listen for changes for the browser context, b=no-bug, c=common, glance

mr-cheffy committed:
Merge branch 'dev' of https://github.com/FlorianButz/desktop into FlorianButz/dev

mr-cheffy committed:
feat: Propagate data into the current context, not the top one, b=no-bug, c=no-component

mr-cheffy committed:
feat: Hint for a restyle after changing boost colors, b=no-bug, c=no-component

FlorianButz committed:
feat: bridge communication using child actor

FlorianButz committed:
feat: Change UI

FlorianButz committed:
fix: missing methods, simplify writing to disk

mr-cheffy committed:
chore: Small fixes, b=no-bug, c=no-component

mr-cheffy committed:
Merge branch 'dev' of https://github.com/FlorianButz/desktop into FlorianButz/dev

mr-cheffy committed:
chore: Clean up, b=no-bug, c=no-component

FlorianButz committed:
fix: change communication of actors, cleanup

mr-cheffy committed:
feat: Prevent flashing when going between pages, b=no-bug, c=common

mr-cheffy committed:
feat: Start taking into account contrast values, b=no-bug, c=no-component

mr-cheffy commented on src/zen/common/ZenUIManager.mjs:
```diff
@@ -50,6 +50,14 @@ var gZenUIManager = {
+    window.gBrowser.addProgressListener({
+      onLocationChange: (aWebProgress) => {
+        if (aWebProgress.isTopLevel) {
+          this.onShownTabChange();
```
This would be cleaner to do inside ZenSiteDataPanel.sys.mjs

FlorianButz replied:
I moved the listener to ZenSiteDataPanel.sys.mjs

mr-cheffy committed:
feat: Correctly calculate color lightness based on contrast, b=no-bug, c=no-component

FlorianButz committed:
feat: update editor UI to look more like Arc's

FlorianButz commented:
<img width="1219" height="758" alt="Bildschirmfoto 2025-11-07 um 8 42 39 PM" src="https://github.com/user-attachments/assets/d6549e3d-b0e1-4c17-a601-cba6a234436f" />
Changed the editor UI to look more like the one from Arc

12th-devs commented:
> <img width="1219" height="758" alt="Bildschirmfoto 2025-11-07 um 8 42 39 PM" src="https://github.com/user-attachments/assets/d6549e3d-b0e1-4c17-a601-cba6a234436f" />
> Changed the editor UI to look more like the one from Arc

@FlorianButz a few questions I had:

1. How does it look in dark mode?

2. How is the size controlled? Drop down list, typing, scroll over the container?

3. What is with the right margin on the size percentage

4. Could the button inside of the gradiant picker be tweaked a bit? It feels a bit out of place in my opinion, but I would love to hear your side.

This all looks so good and I'm really excited, great work man!

12th-devs commented:
> <img width="1219" height="758" alt="Bildschirmfoto 2025-11-07 um 8 42 39 PM" src="https://github.com/user-attachments/assets/d6549e3d-b0e1-4c17-a601-cba6a234436f" />
> Changed the editor UI to look more like the one from Arc

@FlorianButz a few questions I had:

1. How does it look in dark mode?

2. How is the size controlled? Drop down list, typing, scroll over the container?

3. What is with the right margin on the size percentage

4. Could the button inside of the gradiant picker be tweaked a bit? It feels a bit out of place in my opinion, but I would love to hear your side.

This all looks so good and I'm really excited, great work man!

FlorianButz commented:
@12th-devs 
1. There is no dark mode. Arc does not have it, so I left it out for the time being. This can still be added later on if necessary.
2. The size changes the zoom on the website, the dropdown list shows all available system fonts, the typing overrides the text case
3. There are still quite a few visual issues. However to not slow down the development I decided to commit the changes now. Perfecting can be done after everything else is in place.
4. As I said in 3, there's still some issues I currently don't want to prioritize

12th-devs commented:
@FlorianButz that sounds good, thanks for the quick response.

FlorianButz committed:
fix: remove logs, update select properly

mr-cheffy committed:
Merge branch 'dev' into dev

mr-cheffy committed:
chore: Small changes, b=no-bug, c=common

mr-cheffy committed:
Merge branch 'dev' into dev

mr-cheffy committed:
feat: Convert the window into a proper XUL doc and implement scale, b=no-bug, c=no-component

mr-cheffy committed:
feat: Finish migration to xhtml window, b=no-bug, c=common

mr-cheffy committed:
chore: Cleanups, b=no-bug, c=common

FlorianButz committed:
feat: add system font fetching

mr-cheffy committed:
feat: Implemented stylesheet managers, b=no-bug, c=no-component

mr-cheffy committed:
Merge branch 'dev' of https://github.com/FlorianButz/desktop into FlorianButz/dev

FlorianButz committed:
fix: correct color conversion, add stylesheet implementation

mr-cheffy committed:
fix: Fixed crashes on sub-document contexts, b=no-bug, c=no-component

FlorianButz committed:
feat: core zap implementation

FlorianButz committed:
chore: use proper actor communication

greeeeeenbean commented:
Hey all, any update on this?

FlorianButz commented:
@greeeeeenbean Yes, I'm still working on it. I didn't have a lot of time the last few weeks.

The theme of the website can now be synced to the workspace theme.

https://github.com/user-attachments/assets/17012bb7-e589-4ec6-b4b1-3da5f688b673

FlorianButz committed:
feat: workspace color syncing

FlorianButz committed:
chore: Small cleanup

FlorianButz committed:
Merge branch 'dev' of https://github.com/zen-browser/desktop into dev

FlorianButz committed:
chore: update to using isolated includes file

mr-cheffy committed:
feat: Dont render color changes on anonymous documents and run formatter, b=no-bug, c=common

mr-cheffy committed:
feat: Simplify data browsing context fetching, b=no-bug, c=mods

mr-cheffy committed:
feat: Improve detecting anon content, b=no-bug, c=workspaces

mr-cheffy committed:
Merge branch 'dev' into dev

Signed-off-by: mr. m <91018726+mr-cheffy@users.noreply.github.com>

mr-cheffy committed:
chore: Update patches, b=no-bug, c=no-component

FlorianButz committed:
chore: Update boost saving to use JSONFile

FlorianButz committed:
Merge branch 'dev' into dev

Signed-off-by: fen4flo <75260616+FlorianButz@users.noreply.github.com>

Minemetero commented:
Bro reviews every change like he can actually do something about it...

6639835 commented:
> Bro reviews every change like he can actually do something about it...

At least more than you could do 😃

FlorianButz committed:
feat: Add CodeMirror integration for custom css

FlorianButz committed:
Merge branch 'dev' of https://github.com/FlorianButz/desktop into dev

FlorianButz committed:
feat: Account for tabs on right when opening boost window

mr-cheffy commented on prefs/zen/boosts.yaml:
Could you please name it "boosts.yaml"?

mr-cheffy commented on locales/en-US/browser/browser/zen-boosts.ftl:
Could you add the license here?

```
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

```

mr-cheffy commented on src/browser/themes/shared/zen-icons/icons.css:
```diff
@@ -34,6 +34,19 @@
+#PanelUI-zen-gradient-generator-color-remove,
+#zen-gradient-generator-color-remove {
+  list-style-image: url('unpin.svg') !important;
 #appMenu-quit-button2 {
```
These seems to be something wrong going on here?

FlorianButz replied:
I accidentally deleted two lines, it is resolved now.

FlorianButz committed:
Merge branch 'zen-browser:dev' into dev

FlorianButz committed:
chore: Fix syntax error, Rename file, Add license

Remove unnecessary logs from ZenBoostsManager.sys.mjs

FlorianButz committed:
chore: Fix icon location

FlorianButz committed:
feat: Fix color picker knob, Add animation

FlorianButz committed:
feat: Update zap mode

FlorianButz commented:
Updated the zap mode UI. The buttons on the bottom control bar are temporary

https://github.com/user-attachments/assets/b0536b7e-609f-4517-baf7-7a5c9b82c517

FlorianButz committed:
chore: Simplify temporary unzap logic

FlorianButz committed:
feat: Add dissolve effect on zap

mr-cheffy commented on prefs/zen/boosts.yaml:
```diff
@@ -0,0 +1,8 @@
+
+- name: zen.boosts.enabled
+  value: false
+- name: zen.boosts.dissolve-on-zap
```
Nit: add a new line between prefs

mr-cheffy commented on src/browser/themes/shared/zen-icons/common/selectable/brush.svg:
Still under `selectable` folder. `selectable` is used for folder / tab icons btw, not for UI icons.

mr-cheffy commented on src/browser/themes/shared/zen-icons/icons.css:
```diff
@@ -34,6 +34,21 @@
+  list-style-image: url('unpin.svg') !important;
+}
+
 #appMenu-quit-button2 {
```
You can remove this `appMenu-quit-button2` now as its already added at the top?

mr-cheffy committed:
Merge branch 'dev' into dev

mr-cheffy commented on src/browser/themes/shared/zen-icons/icons.css:
```diff
@@ -497,13 +514,34 @@
 }
 
-.geo-icon.blocked-permission-icon {
+.geo-icon.clpermission-icon {
```
Revert this change

mr-cheffy commented on src/zen/boosts/actors/ZenBoostsChild.sys.mjs:
```diff
@@ -0,0 +1,432 @@
+  /**
+   * Called when the actor is destroyed. Cleans up the events.
+   */
+  didDestroy() {
```
Wouldn't it be better to simply listen for "unload"? Not sure window is available at this stage

FlorianButz replied:
Haven't had issues or errors, but listening for unload would probably make more sense since we're only removing event listeners.

mr-cheffy commented on src/zen/boosts/actors/ZenBoostsChild.sys.mjs:
```diff
@@ -0,0 +1,432 @@
+    if (
+      [...ZenBoostsChild.OVERLAY_EVENTS, ...ZenBoostsChild.PREVENTABLE_EVENTS].includes(event.type)
+    ) {
+      this.#overlay.handleEvent(event, ZenBoostsChild.PREVENTABLE_EVENTS.includes(event.type));
```
This looks like it could be optimized a bit, because doing constant checks might be bad for performance

FlorianButz replied:
I changed it to use statically cached Sets, which reduces per-call memory allocation and does a more performant lookup than the linear search an array would give. It should be faster now

mr-cheffy commented on src/zen/boosts/actors/ZenBoostsChild.sys.mjs:
```diff
@@ -0,0 +1,432 @@
+   * Removes the event listeners from the document
+   */
+  #removeEventListeners() {
+    // this.contentWindow.removeEventListener("unload", this._disableZapMode);
```
Remove comment

mr-cheffy commented on src/zen/boosts/actors/ZenBoostsChild.sys.mjs:
```diff
@@ -0,0 +1,432 @@
+    this._handleZapEvent = this.handleZapEvent.bind(this);
+
+    this._disableZapMode = this.disableZapMode.bind();
+    // this.contentWindow.addEventlistener("unload", this._disableZapMode);
```
Remove comment

mr-cheffy commented on src/zen/boosts/actors/ZenBoostsChild.sys.mjs:
```diff
@@ -0,0 +1,432 @@
+    const domain = this.browsingContext.topWindow?.location?.host;
+    if (!domain) return null;
+
+    return this.sendQuery("ZenBoost:GetBoostForDomain", domain);
```
Why's this function async?

mr-cheffy replied:
Also, this could be cached maybe?

FlorianButz replied:
I don't think caching it would make much sense since it is already cached in the `ZenBoostStyles.sys.mjs`. 
I don't know why it's async, I probably wanted to add an await call for the sendQuery but forgot it. I removed the async.

mr-cheffy commented on src/zen/boosts/actors/ZenBoostsChild.sys.mjs:
```diff
@@ -0,0 +1,432 @@
+  async getWebsiteStyle() {
+    const domain = this.browsingContext.topWindow?.location?.host;
+    if (!domain) return null;
+    const styleSheet = (await this.sendQuery("ZenBoost:GetStyleForDomain", domain)).styleSheet; 
```
Instead of repeating `sendQuery`, you can just use getWebsiteBoost()

FlorianButz replied:
I used to update the stylesheet separately for temporarily displaying the zap elements on hover, but since that's gone there's no need for this function anymore.

I removed it.

mr-cheffy commented on src/zen/boosts/actors/ZenBoostsChild.sys.mjs:
```diff
@@ -0,0 +1,432 @@
+    }
+
+    const domain = browsingContext.topWindow?.location?.host;
+    if (!domain) return null;
```
Is `domain` needed here? Could just check for `!boost`

FlorianButz replied:
Correct, it's technically already doing that when checking if the boost is valid before applying the changes.

I removed it.

mr-cheffy reviewed (changes requested):
Im not finished, but this is a good start

mr-cheffy committed:
Merge branch 'dev' into dev

mr-cheffy commented on src/zen/boosts/actors/ZenBoostsParent.sys.mjs:
```diff
@@ -0,0 +1,117 @@
+    super();
+
+    this._observe = this.observe.bind(this);
+    Services.obs.addObserver(this._observe, "zen-boosts-update");
```
No need to call `this._observe` as `Services.obs.addObserver(this)` should already call the `observe` function

mr-cheffy commented on src/browser/themes/shared/zen-icons/common/selectable/brush.svg:
Why remove?

FlorianButz replied:
Old icon, it was replaced by the `paintbrush.svg` and `paintbrush-fill.svg` from the Nucleo icon pack.

mr-cheffy commented on src/zen/boosts/zen-advanced-color-options.css:
```diff
@@ -0,0 +1,87 @@
+}
+
+#zen-boost-advanced-color-options-container {
+  display: flex !important;
```
`!important`?

FlorianButz replied:
Removed the important and the display line entirely, added the flex instruction in the xhtml instead.

mr-cheffy commented on src/zen/boosts/zen-boost-editor.xhtml:
```diff
@@ -0,0 +1,118 @@
+        <button data-l10n="zen-boost-disable" id="zen-boost-disable" class="subviewbutton mod-button small"></button>
+      </hbox>
+
+      <div id="zen-boost-font-wrapper">
```
html elements should be marked with the html namepsace, e.g. `<html:div ...`

mr-cheffy commented on src/zen/boosts/zen-boosts.css:
```diff
@@ -0,0 +1,604 @@
+}
+
+#zen-boost-zap[hideicon="true"] {
+  list-style: none !important;
```
Its bad practice to use `!important`, try to use it as little as possible, it shouldn't be needed here

FlorianButz replied:
Yes, I overused it in the editor css to simplify testing. I'll be removing as many as I can.

mr-cheffy commented on src/zen/boosts/ZenBoostsManager.sys.mjs:
```diff
@@ -0,0 +1,309 @@
+  #file = null;
+
+  constructor() {
+    this.#init();
```
Can just have the contents of `init` here if its just a one-liner

mr-cheffy commented on src/zen/boosts/ZenBoostStyles.sys.mjs:
```diff
@@ -0,0 +1,115 @@
+   */
+  #convertStyleToDataUri(rawStyle) {
+    const encodedStyle = encodeURIComponent(rawStyle);
+    return Services.io.newURI(`data:text/css;charset=utf-8,${encodedStyle}`);
```
Not too sure of having styles applied as `url`s because it might not get cached, but i'd have to check on that

mr-cheffy replied:
Yeah, we might need to write these to a file and give out the file name?

mr-cheffy commented on src/zen/boosts/ZenZapDissolve.sys.mjs:
```diff
@@ -0,0 +1,532 @@
+  `;
+
+  VERT = `
+  precision highp float;
```
Can't this be stored on a separate file?

FlorianButz replied:
Initially wanted to do that, but I wasn't too sure how I should load it in. Having it in class also saves performance as the class itself is created often. 
If the performance impact is too big, which I haven't tested but also didn't notice, I will probably have to create a pool of cached ones anyway.

mr-cheffy commented on src/zen/boosts/ZenZapOverlayChild.sys.mjs:
```diff
@@ -0,0 +1,684 @@
+    this.removeHighlight();
+    this.#resetHoverDiv();
+
+    const { setTimeout } = ChromeUtils.importESModule("resource://gre/modules/Timer.sys.mjs");
```
Import it lazylly at the top

mr-cheffy commented on src/zen/boosts/ZenZapOverlayChild.sys.mjs:
```diff
@@ -0,0 +1,684 @@
+    setTimeout(() => {
+      const useDissolve = Services.prefs.getBoolPref("zen.boosts.dissolve-on-zap");
+      if (!this.window.gReduceMotion && useDissolve) {
+        const elements = this.document.querySelectorAll(cssPath);
```
Does `gReduceMotion` actually exist on `this.window`?

FlorianButz replied:
It does, I haven't found another way of getting to that value. Just using `gReduceMotion` is not possible here

mr-cheffy commented on src/zen/boosts/ZenZapOverlayChild.sys.mjs:
```diff
@@ -0,0 +1,684 @@
+      this.zenBoostsChild.addZapSelector(cssPath);
+      this.#setState(ZapOverlay.STATES.SELECTING);
+      this.onZapUpdate();
+    }, 5);
```
Why `5`?

FlorianButz replied:
There has to be a delay in order for the highlight changes to apply. Five is an arbitrary number that worked for me while testing.

This should probably be replaced but I am not sure how to check for when the style changes are applied as 1 seemed to not be a large enough delay for the changes to apply.

FlorianButz replied:
The magic number and timeout have been replaced with `window.requestAnimationFrame()`. This approach is more reliable

mr-cheffy commented on src/zen/boosts/ZenZapOverlayChild.sys.mjs:
```diff
@@ -0,0 +1,684 @@
+
+    if (boostData.zapSelectors.length == 0)
+      zapList.innerHTML += '<p class="pcenter">Click elements on the page to <b>Zap</b> them</p>';
+    else zapList.innerHTML += "<p>← Click to Unzap</p>";
```
Untranslated strings

mr-cheffy commented on src/zen/boosts/ZenZapOverlayChild.sys.mjs:
```diff
@@ -0,0 +1,684 @@
+    setTimeout(() => {
+      const selection = this.document.querySelectorAll(selector);
+      if (selection.length != 0) this.showHightlight(selection);
+    }, 10);
```
Timeout with un-explained milliseconds

FlorianButz replied:
Same as above, I'll be looking into how to properly wait for the style update instead of waiting an arbitrary amount of time

mr-cheffy commented on src/zen/boosts/ZenZapOverlayChild.sys.mjs:
```diff
@@ -0,0 +1,684 @@
+    const selector = button.getAttribute("selector");
+    this.zenBoostsChild.tempShowZappedElement(selector);
+
+    const { setTimeout } = ChromeUtils.importESModule("resource://gre/modules/Timer.sys.mjs");
```
Import lazily

mr-cheffy commented on src/zen/boosts/ZenZapOverlayChild.sys.mjs:
```diff
@@ -0,0 +1,684 @@
+    }
+
+    this.getElementById("highlight-shadow").display = "initial";
+  }
```
Shouldn't it be `.style.display`?

mr-cheffy commented on src/zen/boosts/ZenZapOverlayChild.sys.mjs:
```diff
@@ -0,0 +1,684 @@
+
+    // Clear all children elements
+    highlightContainerDiv.innerHTML = "";
+    this.getElementById("highlight-shadow").display = "none";
```
Same here

mr-cheffy commented on src/zen/boosts/ZenZapOverlayChild.sys.mjs:
```diff
@@ -0,0 +1,684 @@
+    if (this.#lastOverElement === event.target) return;
+    if (!isZapContent) this.#lastOverElement = event.target;
+
+    if (isZapContent) this.#hideHoverDiv();
```
Could just call `else` from the if statement above

mr-cheffy commented on src/zen/boosts/ZenZapOverlayChild.sys.mjs:
```diff
@@ -0,0 +1,684 @@
+   * @param {Event} event Mouse enter event params
+   */
+  #handleHoverDelegation(event) {
+    if (event.originalTarget.id == "unzap") this.#unzapButtonHover(event);
```
I'd simply add the event listener directly to the button to prevent any possible misunderstandings and call the event less often.

FlorianButz replied:
The reason I did it this way is because registering the event listener directly did nothing. I am not sure why, maybe it was an issue at the time. I'll try to debug and simplify.

mr-cheffy reviewed (changes requested):
Looks good! Please also run `npm run lint:fix`

mr-cheffy committed:
feat: Mark some variables as constant, b=no-bug, c=no-component

mr-cheffy committed:
Merge branch 'dev' of https://github.com/FlorianButz/desktop into FlorianButz/dev

FlorianButz committed:
chore: Resolve requested changes

FlorianButz committed:
chore: Cleanup stylesheet, Remove unecessary !important

FlorianButz committed:
chore: Run lint:fix

FlorianButz committed:
Merge branch 'zen-browser:dev' into dev

FlorianButz committed:
chore: Improve available screen calculation when opening editor

FlorianButz committed:
chore: Fix observers in boosts parent actor

FlorianButz committed:
feat: Improve code editor, Add localizations

FlorianButz committed:
feat: Platform specific style adjustments

FlorianButz committed:
feat: Improve unzap buttons

FlorianButz committed:
chore: Update CSS editor icons to use nucleo icons

FlorianButz committed:
chore: Fix styling issues, Fix using system font

FlorianButz committed:
chore: Fix SelectorComponent positioning in zap mode

FlorianButz committed:
feat: Add exporting / importing boost functionality

FlorianButz committed:
chore: Revert deleting brush.svg

mr-cheffy commented on src/zen/boosts/ZenBoostsManager.sys.mjs:
```diff
@@ -289,6 +287,90 @@ class nsZenBoostsManager {
+    fp.appendFilters(nsIFilePicker.filterAll);
+
+    return new Promise((resolve) => {
+      fp.open(async (result) => {
```
I would just do `fp.open(async (result) => {` also, the function is async for now reason?

FlorianButz replied:
The promise is done so the editor has a callback for when the dialog action is complete. I forgot to remove the async when changing the pattern to promises. Will be removed

FlorianButz committed:
chore: Remove editor window animation

FlorianButz committed:
chore: Improve readability for boost import/export

FlorianButz committed:
feat: Bring back boost names

FlorianButz committed:
chore: Fix zen content check in SelectorComponent

FlorianButz committed:
chore: Fix CSS editor open direction on multi monitor setup

FlorianButz committed:
feat: Add context menu for boost actions

FlorianButz committed:
chore: Fix actor communication, Fix TabSwitch close lifecycle

mr-cheffy commented on prefs/zen/boosts.yaml:
```diff
@@ -0,0 +1,9 @@
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+- name: zen.boosts.enabled
+  value: false
```
Should be `value: "@IS_TWILIGHT@"`

mr-cheffy commented on src/zen/boosts/moz.build:
```diff
@@ -0,0 +1,36 @@
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+EXTRA_JS_MODULES += [
```
I'd do `EXTRA_JS_MODULES.zen.boosts += ...`

mr-cheffy commented on src/zen/boosts/zen-advanced-color-options.css:
```diff
@@ -0,0 +1,86 @@
+
+  @media (-moz-platform: macos) {
+    &[animate="open"] {
+      animation: var(--boost-spring-duration) zen-color-options-panel-animation-macos var(--boost-spring-easing) forwards !important;
```
I dont think we need such complex easing function, no? Id just use the existing zen animation

mr-cheffy commented on src/zen/boosts/zen-selector.css:
```diff
@@ -0,0 +1,232 @@
+      );
+      --boost-spring-duration: 0.6s;
+
+      animation: var(--boost-spring-duration) select-component-in var(--boost-spring-easing) forwards;
```
Same here, overly complex easing function

mr-cheffy commented on src/zen/boosts/zen-boosts.css:
```diff
@@ -0,0 +1,806 @@
+    border-radius: 12px;
+  }
+
+  background-color: #ebebed;
```
Why not just keeping it transparent?

FlorianButz replied:
What do you mean exactly? The button background?

FlorianButz committed:
feat: Refactor to multiple boost per domain system

FlorianButz committed:
chore: Run lint:fix

FlorianButz committed:
feat: Add mica support for the editor, Update close icon

FlorianButz committed:
chore: Adjust styling to be responsive

6639835 reviewed (changes requested):
P1: resetBoost calls gZenBoostsManager.getEmptyBoost(domain) at src/zen/boosts/ZenBoostsEditor.mjs:1067, but the manager only defines getEmptyBoostEntry at src/zen/boosts/ZenBoostsManager.sys.mjs:97. Triggering Reset can throw at runtime. Please use the existing empty-boost API and the correct payload shape, for example getEmptyBoostEntry().boostData.

FlorianButz committed:
chore: Fix resetBoost type error

FlorianButz committed:
chore: Moving js modules to .zen.boosts

FlorianButz committed:
Merge branch 'dev' into dev

Signed-off-by: fen4flo <75260616+FlorianButz@users.noreply.github.com>

FlorianButz committed:
chore: Optimize dissolve effect, Fix zap mode breaking bug

FlorianButz committed:
chore: Run lint:fix

FlorianButz committed:
feat: Improve selector generation

FlorianButz committed:
feat: Add tests for the selector generator

FlorianButz committed:
chore: Fix boost tests

mr-cheffy committed:
feat: Implement smart invert, b=no-bug, c=no-component

mr-cheffy committed:
Merge branch 'dev' of https://github.com/FlorianButz/desktop into FlorianButz/dev

mr-cheffy committed:
Merge branch 'dev' into dev

Signed-off-by: mr. m <91018726+mr-cheffy@users.noreply.github.com>

6639835 commented on src/zen/boosts/actors/ZenBoostsChild.sys.mjs:
```diff
@@ -0,0 +1,451 @@
+        return;
+      }
+    }
+    browsingContext.zenBoostsData = 0;
```
I think the no-boost path only sets `browsingContext.zenBoostsData = 0` and never clears `browsingContext.isZenBoostsInverted`.

It seems like we should explicitly reset `browsingContext.isZenBoostsInverted` in the no-boost path to avoid leaking state across navigations.

FlorianButz replied:
The `isZenBoostsInverted` has been excluded from the color boosting behavior.
The user should be able to disable color boosting and still apply an inversion filter.

6639835 commented on src/zen/boosts/ZenSelectorComponent.sys.mjs:
```diff
@@ -0,0 +1,687 @@
+        traverse(element.parentNode, 0, parentExactElement);
+      }
+
+      const pathParentElementBuilt = build(parentExactElement);
```
In SelectorComponent.getSelectionPath, findBestExactSelector calls document.querySelectorAll(pathParentElementBuilt) even when pathParentElementBuilt is empty

FlorianButz committed:
chore: Fix lint errors

FlorianButz committed:
chore: Fix close button padding on mac

mr-cheffy committed:
Merge branch 'dev' into dev

mr-cheffy commented on src/zen/boosts/actors/ZenBoostsChild.sys.mjs:
```diff
@@ -0,0 +1,493 @@
+        return this.#currentState === ZenBoostsChild.STATES.ZAP;
+      case "ZenBoost:SelectorPickerModeEnabled":
+        return this.#currentState === ZenBoostsChild.STATES.PICKER;
+      case "ZenBoost:OpenInspector":
```
Why do we need this exactly? Doesn't make much more sense to send the message as we recieve it. Doesn't it make an infinite loop?

FlorianButz replied:
These things report the state to the editor which is needed for the button color. Why would there be an infinite loop?

mr-cheffy replied:
Im talking about `ZenBoost:OpenInspector`

FlorianButz replied:
No, messages received from the outside will land in the child actor. However the child actor can not open the inspector as it does not have the required references. So it has to send the message back to the Parent, which does not create a recursive loop

mr-cheffy commented on src/zen/boosts/actors/ZenBoostsChild.sys.mjs:
```diff
@@ -0,0 +1,493 @@
+        this.#loadStyleSheet(boost.styleSheet);
+      }
+
+      browsingContext.isZenBoostsInverted = boostData.smartInvert;
```
We are not resetting the invert if boost doesn't exist

mr-cheffy commented on src/zen/boosts/actors/ZenBoostsChild.sys.mjs:
```diff
@@ -0,0 +1,493 @@
+          0.1 + boostData.dotDistance * 0.6 * boostData.brightness
+        );
+
+        let primaryGradientColor = boost.workspaceGradient[0].c ?? this.#rgbToHsl([75, 75, 75]);
```
Directly use the result of `this.#rgbToHsl([75, 75, 75])` as a constant to avoid excessive computations. Also, what if there are no workspace gradients?

mr-cheffy commented on src/zen/boosts/actors/ZenBoostsParent.sys.mjs:
```diff
@@ -0,0 +1,154 @@
+   * Called when the actor is destroyed. Cleans up the observer.
+   */
+  async destroy() {
+    super.destroy();
```
`await`

mr-cheffy commented on src/zen/boosts/actors/ZenBoostsParent.sys.mjs:
```diff
@@ -0,0 +1,154 @@
+
+        let toolbox = gDevTools.getToolboxForTab(tab);
+
+        if (toolbox) {
```
It shouldn't matter, no? We always want to open the "inspector" tab

FlorianButz replied:
No, if we don't do this clicking the button multiple times will open a new inspector every time. We have to disable it if there's an open one.

mr-cheffy commented on src/zen/boosts/actors/ZenBoostsParent.sys.mjs:
```diff
@@ -0,0 +1,154 @@
+        if (!data.selector) {
+          break;
+        }
+        if (!data.domain) {
```
Can be simplified to `!data.action && ...`

mr-cheffy commented on src/zen/boosts/nsZenBoostsBackend.cpp:
```diff
@@ -0,0 +1,257 @@
+  if (!aPresContext) {
+    return;
+  }
+  mCurrentPresContext = aPresContext;
```
Maybe I should be using a weak ref here

mr-cheffy commented on src/zen/boosts/zen-advanced-color-options.css:
```diff
@@ -0,0 +1,52 @@
+
+  @media (-moz-platform: macos) {
+    &[animate="open"] {
+      animation: 0.4 zen-color-options-panel-animation-macos cubic-bezier(0.14, 1.43, 0.56, 1.01) forwards !important;
```
We don't need this anymore as we have native popups now

mr-cheffy commented on src/zen/boosts/zen-boost-editor.xhtml:
```diff
@@ -0,0 +1,152 @@
+  <link rel="stylesheet" href="chrome://browser/skin/zen-icons/icons.css" />
+  <link rel="stylesheet" href="chrome://browser/content/zen-styles/zen-advanced-color-options.css" />
+
+  <link rel="localization" href="browser/zen-boosts.ftl"/>
```
Wrap in `<linkset>` and mark it as html `<html:link ...`

mr-cheffy commented on src/zen/boosts/zen-boost-editor.xhtml:
```diff
@@ -0,0 +1,152 @@
+    </hbox>
+  </vbox>
+
+  <popupset id="mainPopupSet">
```
Move to top

mr-cheffy commented on src/zen/boosts/zen-boost-editor.xhtml:
```diff
@@ -0,0 +1,152 @@
+    </panel>
+  </popupset>
+
+  <menupopup id="zenBoostContextMenu">
```
Move to top

mr-cheffy commented on src/zen/boosts/zen-boosts.css:
```diff
@@ -0,0 +1,804 @@
+#zenBoostWindow {
+  /* For the mica effect we want a white tint */
+  color-scheme: light !important;
+  appearance: none !important;
```
Why important?

FlorianButz replied:
This didn't work without important if I recall correctly. I will check again though, maybe I had other styles conflicting

mr-cheffy commented on src/zen/boosts/zen-boosts.css:
```diff
@@ -0,0 +1,804 @@
+}
+
+#zen-boost-code-editor-root {
+  flex: 1;
```
Can just give it `flex="1"` to the element

mr-cheffy commented on src/zen/boosts/ZenZapDissolve.sys.mjs:
```diff
@@ -0,0 +1,591 @@
+  get markup() {
+    return `
+    <template>
+      <link rel="stylesheet" href="chrome://browser/content/zen-styles/content/zen-zap.css" />
```
Shouldn't it be `scoped`?

mr-cheffy commented on src/zen/boosts/ZenZapOverlayChild.sys.mjs:
```diff
@@ -0,0 +1,383 @@
+   * Removes all event listeners and removes the overlay from the Anonymous Content
+   */
+  tearDown() {
+    this.#selectorComponent.tearDown();
```
Might also need to clear `this.window` and document

mr-cheffy commented on src/zen/common/styles/zen-animations.css:
```diff
@@ -4,6 +4,29 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
 
+@keyframes zen-color-options-panel-animation-macos {
```
Not needed anymore

mr-cheffy commented on src/zen/common/styles/zen-animations.css:
```diff
@@ -16,6 +39,18 @@
   }
 }
 
+@keyframes zen-boost-color-picker-dot-animation {
```
I wouldn't use this either, let it be static, less motion

mr-cheffy reviewed (changes requested):
We are so close to finishing this!

FlorianButz committed:
chore: Fix undefined errors

FlorianButz committed:
chore: Fix type error when opening boost editor

FlorianButz committed:
feat: Use of prompt instead of inline textfield for renaming boosts

FlorianButz committed:
chore: Rename gradient generator update to be generic

FlorianButz committed:
chore: Fix minor issues that broke everything

FlorianButz committed:
chore: Fix ZenZapOverlayChild not setting text content properly

FlorianButz committed:
chore: Fix lint errors

FlorianButz committed:
chore: Remove unused animation, Reset smart invert when boost null

FlorianButz committed:
chore: Fix not clearing ref to window and doc

FlorianButz committed:
chore: Fix bug when cleaning up dissolve effect

FlorianButz commented on src/zen/workspaces/ZenGradientGenerator.mjs:
```diff
@@ -1675,8 +1678,8 @@ export class nsZenThemePicker extends nsZenMultiWindowFeature {
     let currentWorkspace = gZenWorkspaces.getActiveWorkspace();
 
+    currentWorkspace.theme = gradient;
     if (!skipSave) {
-      currentWorkspace.theme = gradient;
       gZenWorkspaces.saveWorkspace(currentWorkspace);
     }
```
This does not regard this PR and therefor won't be fixed here.

FlorianButz committed:
chore: Remove dot animation because cheff said so

FlorianButz committed:
chore: Fix not sanitizing HTML

mr-cheffy committed:
Merge branch 'dev' into dev

mr-cheffy committed:
fix: Fixed GTK windows, b=no-bug, c=no-component

mr-cheffy committed:
Merge branch 'dev' of https://github.com/FlorianButz/desktop into FlorianButz/dev

FlorianButz committed:
chore: Fix not opening boost editor as dialog

FlorianButz committed:
chore: Change default contrast value to be lower

mr-cheffy committed:
feat: Fixed animations not colouring correctly, b=no-bug, c=no-component

FlorianButz committed:
Merge branch 'dev' of https://github.com/FlorianButz/desktop into dev

FlorianButz committed:
chore: Update child actor to use windowGlobalChild instead

FlorianButz committed:
Merge branch 'dev' into dev

Signed-off-by: fen4flo <75260616+FlorianButz@users.noreply.github.com>

FlorianButz committed:
chore: Change to storing normalized dot pos

FlorianButz committed:
chore: Fix clamp function in backend

FlorianButz committed:
feat: Separate boost data and user css

mr-cheffy committed:
Merge branch 'dev' into dev

mr-cheffy committed:
feat: Start deducing colors from browsing context, not pres shells, b=no-bug, c=no-component

mr-cheffy committed:
feat: Small changes to the UI, b=no-bug, c=no-component
